app/src/gecko/java/org/mozilla/tv/firefox/ext/EngineView.kt

```diff
@@ -38,13 +45,103 @@ fun EngineView.setupForApp() {
     }
 }
 
+@VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
 fun EngineView.evalJS(javascript: String, callback: ValueCallback<String>? = null) {
     geckoView?.session?.loadUri("javascript:$javascript")
     callback.forceExhaustive
 }
 
 fun EngineView.pauseAllVideoPlaybacks() {
-    evalJS("document.querySelectorAll('video').forEach(v => v.pause());")
+    evalJS(pauseVideo)
+}
+
+fun EngineView.domElementCache() {
+    evalJS(CACHE_JS)
+}
+
+fun EngineView.domElementRestore() {
+    evalJS("if ($CACHE_VAR) $CACHE_VAR.focus();")
+}
+
+fun EngineView.observePlaybackState() {
+    evalJS(JS_OBSERVE_PLAYBACK_STATE)
+}
+
+private fun EngineView.evalJSWithTargetVideo(getExpressionToEval: (videoId: String) -> String) {
+    val ID_TARGET_VIDEO = "targetVideo"
+    val GET_TARGET_VIDEO_OR_RETURN = """
+            |var videos = Array.from(document.querySelectorAll('video'));
+            |if (videos.length === 0) { return; }
+            |
+            |var $ID_TARGET_VIDEO = videos.find(function (video) { return !video.paused });
+            |if (!$ID_TARGET_VIDEO) {
+            |    $ID_TARGET_VIDEO = videos[0];
+            |}
+            """.trimMargin()
+
+    val expressionToEval = getExpressionToEval(ID_TARGET_VIDEO)
+    evalJS("""
+            |(function() {
+            |    $GET_TARGET_VIDEO_OR_RETURN
+            |    $expressionToEval
+            |})();
+            """.trimMargin())
```

Comment:
Correct me if I'm wrong, but moving JS code into the `JS` object helps prevent repetition between System and Gecko variants, right?  I wonder if these larger chunks of code should live there as well.  It seems like they should both live in the same place, wherever that is.